zookeeper-server/src/test/java/org/apache/zookeeper/EnforceAuthenticationTest.java

```diff
@@ -0,0 +1,274 @@
+  }
+
+  @Test
+  public void testEnforceAuthenticationNewBehaviourWithNetty() throws Exception {
+    Map<String, String> prop = new HashMap<>();
+    prop.put(removeZooKeeper(AuthenticationHelper.ENFORCE_AUTH_ENABLED), "true");
+    prop.put(removeZooKeeper(AuthenticationHelper.ENFORCE_AUTH_SCHEME), "digest");
+    prop.put("serverCnxnFactory", "org.apache.zookeeper.server.NettyServerCnxnFactory");
+    startServer(prop);
+    testEnforceAuthNewBehaviour(true);
+  }
+
+  /**
+   * Client operations are allowed only after the authentication is done
+   */
+  private void testEnforceAuthNewBehaviour(boolean netty)
+      throws Exception {
+    ZKClientConfig config = new ZKClientConfig();
+    if (netty) {
+      config.setProperty(ZKClientConfig.ZOOKEEPER_CLIENT_CNXN_SOCKET,
+          "org.apache.zookeeper.ClientCnxnSocketNetty");
+    }
+    CountDownLatch countDownLatch = new CountDownLatch(1);
+    ZooKeeper client =
+        new ZooKeeper("127.0.0.1:" + clientPort, CONNECTION_TIMEOUT, getWatcher(countDownLatch),
+            config);
+    countDownLatch.await();
+    String path = "/newAuth" + System.currentTimeMillis();
+    String data = "someData";
+
+    //try without authentication
+    try {
+      client.create(path, data.getBytes(), Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+      fail("SessionClosedRequireAuthException is expected.");
+    } catch (KeeperException.SessionClosedRequireAuthException e) {
+      //do nothing
+    }
+    client.close();
+    countDownLatch = new CountDownLatch(1);
+    client =
+        new ZooKeeper("127.0.0.1:" + clientPort, CONNECTION_TIMEOUT, getWatcher(countDownLatch),
+            config);
+    countDownLatch.await();
+
+    // try operations after authentication
+    String idPassword = "user1:pass1";
+    client.addAuthInfo("digest", idPassword.getBytes());
+    client.create(path, data.getBytes(), Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+    byte[] data1 = client.getData(path, false, null);
+    assertEquals(data, new String(data1));
+    client.close();
+  }
+
+  private String removeZooKeeper(String prop) {
+    return prop.replace("zookeeper.", "");
+  }
+
+  private File getConfFile(Map<String, String> additionalProp) throws IOException {
+    clientPort = PortAssignment.unique();
+    StringBuilder sb = new StringBuilder();
+    sb.append("standaloneEnabled=true" + "\n");
+    if (null != additionalProp) {
+      for (Map.Entry<String, String> entry : additionalProp.entrySet()) {
+        sb.append(entry.getKey());
+        sb.append("=");
+        sb.append(entry.getValue());
+        sb.append("\n");
+      }
+    }
+    String currentQuorumCfgSection = sb.toString();
+    return new MainThread(1, clientPort, currentQuorumCfgSection, false).getConfFile();
+  }
+
+  private void startServer(Map<String, String> additionalProp) throws IOException {
+    clientPort = PortAssignment.unique();
+    StringBuilder sb = new StringBuilder();
+    sb.append("standaloneEnabled=true");
```

Comment:
StandAlone mode is enabled by default. Do you think it would be possible to reuse `launchServers()` method in QuorumPeerTestBase instead of implementing it here?

Reply:
yes, it can be used.
Used LaunchServers() after adding new overloaded APIs to launch servers